container_api/src/container.rs

```diff
@@ -8,59 +7,93 @@ use Holochain;
 
 use holochain_core::{logger::Logger, persister::SimplePersister};
 use holochain_core_types::agent::Agent;
+use holochain_net::p2p_network::P2pNetwork;
 use std::{
+    clone::Clone,
     collections::HashMap,
     convert::TryFrom,
     fs::File,
     io::prelude::*,
     sync::{Arc, Mutex, RwLock},
+    thread,
 };
 
-use holochain_net::p2p_network::P2pNetwork;
+use config::{Configuration, InterfaceConfiguration, InterfaceDriver, StorageConfiguration};
+use interface::{self, DispatchRpc, InstanceMap, Interface, RpcDispatcher};
+use interface_impls;
+use jsonrpc::JsonRpcRequest;
 
 /// Main representation of the container.
 /// Holds a `HashMap` of Holochain instances referenced by ID.
-///
+
 /// A primary point in this struct is
-/// ```load_config(&mut self, config: &Configuration) -> Result<(), String>```
+/// `load_config(&mut self, config: &Configuration) -> Result<(), String>`
 /// which takes a `config::Configuration` struct and tries to instantiate all configured instances.
 /// While doing so it has to load DNA files referenced in the configuration.
 /// In order to not bind this code to the assumption that there is a filesystem
 /// and also enable easier testing,
 /// a DnaLoader has to be injected on creation.
 /// This is a closure that returns a Dna object for a given path string.
 pub struct Container {
-    pub instances: HashMap<String, Holochain>,
-    dna_loader: DnaLoader,
+    pub instances: InstanceMap,
+    config: Configuration,
+    interface_threads: HashMap<String, InterfaceThreadHandle>,
+    pub dna_loader: DnaLoader,
 }
 
+type InterfaceThreadHandle = thread::JoinHandle<Result<(), String>>;
 type DnaLoader = Arc<Box<FnMut(&String) -> Result<Dna, HolochainError> + Send>>;
 
 impl Container {
     /// Creates a new instance with the default DnaLoader that actually loads files.
-    pub fn new() -> Self {
+    pub fn with_config(config: Configuration) -> Self {
```

Comment:
```suggestion
    pub fn from_config(config: Configuration) -> Self {
```

Reply:
@lucksus don't we use the `with_` convention for constructors? Why `from_` here?

Reply:
Actually, it should be the `From<T>` trait: https://doc.rust-lang.org/std/convert/trait.From.html

But the config could be broken, which is why I then used TryFrom below. Why do we need a new() at all?